

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data &mdash; SpaCon_tutorials 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=f2a433a1"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data" href="SpaCon_tutorial4_for_marmoset_ST_and_FC.html" />
    <link rel="prev" title="Tutorial 2: SpaCon for layer6 CT neuron subtype classification" href="SpaCon_tutorial2_for_mouse_celltype_classification.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SpaCon_tutorials
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Installation%20Guide%20for%20SpaCon.html">Installation Guide for <code class="docutils literal notranslate"><span class="pre">SpaCon</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial1_for_mouse_3D_ST.html">Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial2_for_mouse_celltype_classification.html">Tutorial 2: SpaCon for layer6 CT neuron subtype classification</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial4_for_marmoset_ST_and_FC.html">Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SpaCon_tutorials</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/SpaCon_tutorial3_for_mouse_widefield_FC_data.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Tutorial-3:-SpaCon-for-mouse-spatial-transcriptomics-and-widefield-functional-connectivity-data">
<h1>Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data<a class="headerlink" href="#Tutorial-3:-SpaCon-for-mouse-spatial-transcriptomics-and-widefield-functional-connectivity-data" title="Link to this heading"></a></h1>
<p>This tutorial demonstrates how to use SpaCon to integrate mouse gene expression and wide‑field calcium imaging data. We first mapped the mouse spatial transcriptomics data onto a cortical atlas and aligned it with the wide‑field calcium imaging data.</p>
<p>The spatial transcriptomics data used here is the MERFISH dataset from the study published in <em>Nature</em>: <a class="reference external" href="https://www.nature.com/articles/s41586-023-06808-9">Molecularly defined and spatially resolved cell atlas of the whole mouse brain</a>. The wide‑field calcium imaging data come from: <a class="reference external" href="https://www.nature.com/articles/s41467-024-47762-y">Diverse and asymmetric patterns of single‑neuron projectome in regulating interhemispheric connectivity</a>.</p>
<p>We co‑registered both datasets to the same spatial resolution. The processed data for this tutorial can be downloaded from this <a class="reference external" href="https://drive.google.com/drive/folders/1lQQQVjXt8lvciuKq_Jkp4LGsR_EQXFAi?usp=sharing">Google Drive link</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import spacon
from spacon.utils import build_spatial_graph, build_connection_graph, neighbor_sample, model_train, model_eval, clustering

from spacon.model import SpaCon

import datetime
import os
import scanpy as sc
import matplotlib.pyplot as plt

import torch
import numpy as np
import random

import warnings
warnings.filterwarnings(&quot;ignore&quot;)
mus = &#39;mouse_3&#39;
if mus == &#39;mouse_1&#39;:  # coronal
    plot_x, plot_y = &#39;z&#39;, &#39;y&#39;
    figsize = (5,5)
elif mus == &#39;mouse_3&#39;:   # sagittal
    plot_x, plot_y = &#39;x&#39;, &#39;y&#39;
    figsize = (11,5)


def set_seed(seed: int):
    os.environ[&#39;PYTHONHASHSEED&#39;] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

set_seed(42)
</pre></div>
</div>
</div>
<p><strong>Data preprocessing</strong></p>
<p><strong>Load spatial transcriptomics data</strong></p>
<p>If there are too many genes (for example, more than 5,000), we recommend first screening for highly variable genes using the following method:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>n_top_genes = 3000
sc.pp.highly_variable_genes(adata, flavor=&quot;seurat&quot;, n_top_genes=n_top_genes)
adata = adata[:, adata.var.highly_variable]
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.read_h5ad(&#39;/mnt/Data16Tc/home/haichao/code/SpaCon/ST_FC_cluster/mouse1/data/zxw1_wide_field/zxw1_cortical_map_half_brain_match_wf_conn.h5ad&#39;)   # gene expression has been normalize_total and log1p
adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 3372 × 1122
    obs: &#39;x&#39;, &#39;y&#39;, &#39;wf_index&#39;
    uns: &#39;log1p&#39;
    obsm: &#39;X_spatial_2d&#39;
</pre></div></div>
</div>
<p><strong>Build spatial graph</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_spatial_graph</span></code> function constructs a spatial graph using the three-dimensional spatial coordinates of the spatial transcriptome. The main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: Spatial transcriptomics data must include the three-dimensional coordinates for each spot (i.e., the slice number and the two-dimensional coordinates within that slice).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">section_order</span></code>: A slice order list where the sequence represents the original arrangement of each slice within the brain. While the slices can be oriented differently, their relative order must be strictly maintained.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rad_cutoff</span></code>: Neighborhood radius, each spot will have edges added to all other spots within its neighborhood radius.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rad_cutoff_Zaxis</span></code>: Inter-slice neighborhood radius, each spot will have edges added to spots in adjacent slices that are within this radius.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sec_x</span></code>: The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the x-coordinate of each spot within its slice.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sec_y</span></code>: The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the y-coordinate of each spot within its slice.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">key_section</span></code>: Column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the slice number (where different numbers indicate different slices).</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># calculate the spatial graph for the adata
ST_graph_data, st_adj = build_spatial_graph(adata=adata,  k_cutoff=15, model=&#39;KNN&#39;,
                                            sec_x=&#39;y&#39;, sec_y=&#39;x&#39;, is_3d=False)
ST_graph_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[3372, 1122], edge_index=[2, 53952])
</pre></div></div>
</div>
<p><strong>Load connectivity data and build connection graph</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_connection_graph</span></code> function uses connection information to construct a three-dimensional connection graph. The main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">nt_adj</span></code>: An n x n two-dimensional matrix, where n is the number of spots in the spatial transcriptomics data, representing the connection strength between spots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">threshold</span></code>: Filtering threshold, connection strengths below this value will be set to zero.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>distance_weight = True
decay_rate = 0.006
neighbor_weight1 = False
neighbor_weight1_percentage = 30

if distance_weight:
    wf_FC_mouse1 = np.load(&#39;/mnt/Data16Tc/home/haichao/code/SpaCon/ST_FC_cluster/mouse1/data/zxw1_wide_field/wf_FC_mouse1_fliter_100um.npy&#39;)
    coor = np.array(adata.obs[[&#39;x&#39;, &#39;y&#39;]])
    for i in range(wf_FC_mouse1.shape[0]):
        distances = np.linalg.norm(coor - coor[i], axis=1)
        neighbor = np.percentile(distances, neighbor_weight1_percentage)

        weight = 1/(np.exp(-decay_rate * distances))
        # weight = weight/np.max(weight)
        if neighbor_weight1:
            weight[distances &lt; neighbor] = 1
        # weight = weight/np.max(weight)
        # print(weight.max())
        wf_FC_mouse1[i] = np.multiply(wf_FC_mouse1[i], weight)
        # break
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>def filter_matrix(mat, thr, per):
    n = mat.shape[0]
    k_per_row = int(per * n)  # Calculate the maximum number of elements to retain per row (150)
    filtered_mat = np.zeros_like(mat)  # Initialize the filtered matrix

    for i in range(n):
        row = mat[i, :].copy()  # Copy the current row to avoid modifying the original matrix

        # Step 1: Retain elements greater than 0.7
        mask = row &gt; thr
        valid_indices = np.where(mask)[0]

        if len(valid_indices) == 0:
            continue  # No matching elements, skip

        # Step 2: Sort in descending order by value and select the top k elements
        valid_values = row[valid_indices]
        sorted_indices = np.argsort(-valid_values)  # Indices for descending sort

        k = min(k_per_row, len(sorted_indices))
        selected = sorted_indices[:k]
        selected_indices = valid_indices[selected]

        # Update the filtered matrix
        filtered_mat[i, selected_indices] = row[selected_indices]

    # Optional step: Maintain matrix symmetry
    # filtered_mat = np.maximum(filtered_mat, filtered_mat.T)

    return filtered_mat

thr = 0.8
max_retention_each_row = 0.1
wf_FC_mouse1 = filter_matrix(wf_FC_mouse1, thr=thr, per=max_retention_each_row)
# for i in range(wf_FC_mouse1.shape[0]):
#     wf_FC_mouse1[i,i] = 2
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>wf_FC_mouse1[wf_FC_mouse1 &lt; thr] = 0
count_after = np.count_nonzero(wf_FC_mouse1)
proportion_after = count_after/(wf_FC_mouse1.shape[0]*wf_FC_mouse1.shape[1])
print(proportion_after)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
0.08214357580183748
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>NT_graph_data = build_connection_graph(adata, wf_FC_mouse1, threshold=thr)
NT_graph_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[3372, 1122], edge_index=[2, 934013])
</pre></div></div>
</div>
<p><strong>Neighbor-based subgraph sampling</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">neighbor_sample</span></code> function performs subgraph sampling from the input spatial graph and connection graph. Its main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_size</span></code>: The batch size for model training.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">train_num_neighbors</span></code>: The number of neighbors to sample for each node in each iteration. This parameter is used by the data loader during the model training process.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">eval_num_neighbors</span></code>: The number of neighbors to sample for each node in each iteration. This parameter is used by the data loader during the model evaluation process. If an entry is set to -1, all neighbors will be included.(default:<code class="docutils literal notranslate"><span class="pre">[-1]</span></code>)</p></li>
</ul>
<p>The function returns three data loaders: <code class="docutils literal notranslate"><span class="pre">train_loader</span></code>, <code class="docutils literal notranslate"><span class="pre">evaluate_loader_con</span></code>, and <code class="docutils literal notranslate"><span class="pre">evaluate_loader_spa</span></code>. The <code class="docutils literal notranslate"><span class="pre">train_loader</span></code> is used during the <strong>model training process</strong>. Meanwhile, <code class="docutils literal notranslate"><span class="pre">evaluate_loader_con</span></code> and <code class="docutils literal notranslate"><span class="pre">evaluate_loader_spa</span></code> are used for <strong>model evaluation</strong> on the <strong>connection graph</strong> and <strong>spatial graph</strong>, respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>train_loader, evaluate_loader_con, evaluate_loader_spa = neighbor_sample(NT_graph_data, ST_graph_data, batch_size=64, train_num_neighbors=[20, 10, 10], num_workers=4)
</pre></div>
</div>
</div>
<p><strong>Model training</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>device = torch.device(&#39;cuda:0&#39; if torch.cuda.is_available() else &#39;cpu&#39;)

# hyper-parameters
num_epoch = 10
lr = 0.0001
weight_decay = 1e-4
hidden_dims = [adata.X.shape[1]] + [256, 128, 32]
# model
# fusion_method indicates the feature fusion method of the middle layer, you can choose &#39;add&#39; or &#39;concat&#39;
model = SpaCon(hidden_dims=hidden_dims, fusion_method=&#39;concat&#39;).to(device)
# if model_save_path=None, the model will not be saved
results_save_path = f&quot;./results_widefield/{str(datetime.datetime.now().strftime(&#39;%Y_%m_%d_%H_%M_%S&#39;))}/&quot;
os.makedirs(results_save_path, exist_ok=True)

model = model_train(num_epoch, lr, weight_decay, model, train_loader, st_adj, model_save_path=results_save_path, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:1|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 26.57it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:2|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 32.05it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:3|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 29.43it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:4|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 30.42it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:5|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 28.31it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:6|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 29.57it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:7|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 30.59it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:8|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 30.99it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:9|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 32.25it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:10|10
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 53/53 [00:01&lt;00:00, 31.54it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>


Training completed! The model parameters have been saved to ./results_widefield/2025_07_24_14_06_20/model_params.pth
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p><strong>Model evaluation</strong></p>
<p>The features obtained after model dimensionality reduction, named <code class="docutils literal notranslate"><span class="pre">feature_spa</span></code> and <code class="docutils literal notranslate"><span class="pre">feature_con</span></code>, are stored in the returned <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code>. These features can be used for subsequent cluster analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = model_eval(model, adata, NT_graph_data, ST_graph_data, evaluate_loader_con, evaluate_loader_spa, st_adj, layer_eval=True, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Evaluating: 100%|██████████| 10116/10116 [00:01&lt;00:00, 6952.61it/s]
Evaluating: 100%|██████████| 10116/10116 [00:01&lt;00:00, 8215.90it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The results have been saved in adata.obsm
AnnData object with n_obs × n_vars = 3372 × 1122
    obs: &#39;x&#39;, &#39;y&#39;, &#39;wf_index&#39;
    uns: &#39;log1p&#39;
    obsm: &#39;X_spatial_2d&#39;, &#39;feature_spa&#39;, &#39;feature_con&#39;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p><strong>Clustering</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">clustering</span></code> function performs clustering using the louvain algorithm, with the following key parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: The AnnData object obtained previously, which contains the clustering features (<code class="docutils literal notranslate"><span class="pre">feature_spa</span></code>, <code class="docutils literal notranslate"><span class="pre">feature_con</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alpha</span></code>: This parameter adjusts the contribution of local spatial information versus global connection information in the clustering results.</p>
<ul>
<li><p>When <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">1</span></code>, the clustering will incorporate more global information.</p></li>
<li><p>When <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">0</span></code>, the clustering will focus more on local information. You can set different <code class="docutils literal notranslate"><span class="pre">alpha</span></code> values based on your downstream tasks.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">adata_save_path</span></code>: The path where the results will be saved.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cluster_resolution</span></code>: The clustering resolution used during the louvain clustering process.</p></li>
</ul>
<p>The returned <code class="docutils literal notranslate"><span class="pre">path</span></code> indicates where the clustering results are saved.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x</th>
      <th>y</th>
      <th>wf_index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>41</td>
      <td>20</td>
      <td>22</td>
    </tr>
    <tr>
      <th>1</th>
      <td>42</td>
      <td>20</td>
      <td>23</td>
    </tr>
    <tr>
      <th>2</th>
      <td>43</td>
      <td>20</td>
      <td>24</td>
    </tr>
    <tr>
      <th>3</th>
      <td>44</td>
      <td>20</td>
      <td>25</td>
    </tr>
    <tr>
      <th>4</th>
      <td>45</td>
      <td>20</td>
      <td>26</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3367</th>
      <td>25</td>
      <td>103</td>
      <td>1025</td>
    </tr>
    <tr>
      <th>3368</th>
      <td>26</td>
      <td>103</td>
      <td>1047</td>
    </tr>
    <tr>
      <th>3369</th>
      <td>27</td>
      <td>103</td>
      <td>1071</td>
    </tr>
    <tr>
      <th>3370</th>
      <td>28</td>
      <td>103</td>
      <td>1096</td>
    </tr>
    <tr>
      <th>3371</th>
      <td>29</td>
      <td>103</td>
      <td>1122</td>
    </tr>
  </tbody>
</table>
<p>3372 rows × 3 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata, path = clustering(adata, alpha=1, adata_save_path=results_save_path, cluster_resolution=0.6)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The clustering results have been saved in ./results_widefield/2025_07_24_14_06_20/feature_add_weight1/Clusters_res0.6/
AnnData object with n_obs × n_vars = 3372 × 1122
    obs: &#39;x&#39;, &#39;y&#39;, &#39;wf_index&#39;, &#39;clusters&#39;
    uns: &#39;log1p&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;clusters&#39;, &#39;clusters_colors&#39;
    obsm: &#39;X_spatial_2d&#39;, &#39;feature_spa&#39;, &#39;feature_con&#39;, &#39;feature_add&#39;, &#39;X_umap&#39;, &#39;spatial&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial3_for_mouse_widefield_FC_data_23_1.png" src="_images/SpaCon_tutorial3_for_mouse_widefield_FC_data_23_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obsm[&#39;spatial&#39;] = adata.obs[[&#39;y&#39;, &#39;x&#39;]].values
sc.pl.spatial(adata, color=&#39;clusters&#39;, spot_size=1, show=False)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;Axes: title={&#39;center&#39;: &#39;clusters&#39;}, xlabel=&#39;spatial1&#39;, ylabel=&#39;spatial2&#39;&gt;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial3_for_mouse_widefield_FC_data_24_1.png" src="_images/SpaCon_tutorial3_for_mouse_widefield_FC_data_24_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>re_label = spacon.utils.refine_label(adata, radius=25, key=&#39;clusters&#39;)
adata.obs[&#39;refine&#39;] = re_label
adata.obs[&#39;refine&#39;] = adata.obs[&#39;refine&#39;].astype(&#39;category&#39;)
sc.pl.spatial(adata, color=&#39;refine&#39;, spot_size=1, show=False)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;Axes: title={&#39;center&#39;: &#39;refine&#39;}, xlabel=&#39;spatial1&#39;, ylabel=&#39;spatial2&#39;&gt;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial3_for_mouse_widefield_FC_data_25_1.png" src="_images/SpaCon_tutorial3_for_mouse_widefield_FC_data_25_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>for c in adata.obs[&#39;clusters&#39;].unique():
    temp_adata = adata[adata.obs[&#39;clusters&#39;] == c]
    plt.figure(figsize=(4,6))
    plt.scatter(adata.obs[&#39;x&#39;].values, adata.obs[&#39;y&#39;].values, c=&#39;#d3d3d3&#39;, s=10)
    plt.scatter(temp_adata.obs[&#39;x&#39;].values, temp_adata.obs[&#39;y&#39;].values, c=&#39;#FF8C00&#39;, s=10)
    # plt.savefig(f&#39;{path}/{c}.png&#39;)
    plt.show()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="SpaCon_tutorial2_for_mouse_celltype_classification.html" class="btn btn-neutral float-left" title="Tutorial 2: SpaCon for layer6 CT neuron subtype classification" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="SpaCon_tutorial4_for_marmoset_ST_and_FC.html" class="btn btn-neutral float-right" title="Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, quhaichao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>