

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity &mdash; SpaCon_tutorials 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=f2a433a1"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 2: SpaCon for layer6 CT neuron subtype classification" href="SpaCon_tutorial2_for_mouse_celltype_classification.html" />
    <link rel="prev" title="Installation Guide for SpaCon" href="Installation%20Guide%20for%20SpaCon.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SpaCon_tutorials
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Installation%20Guide%20for%20SpaCon.html">Installation Guide for <code class="docutils literal notranslate"><span class="pre">SpaCon</span></code></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial2_for_mouse_celltype_classification.html">Tutorial 2: SpaCon for layer6 CT neuron subtype classification</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial3_for_mouse_widefield_FC_data.html">Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial4_for_marmoset_ST_and_FC.html">Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SpaCon_tutorials</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/SpaCon_tutorial1_for_mouse_3D_ST.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Tutorial-1:-SpaCon-for-mouse-3D-spatial-transcriptome-and-structural-connectivity">
<h1>Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity<a class="headerlink" href="#Tutorial-1:-SpaCon-for-mouse-3D-spatial-transcriptome-and-structural-connectivity" title="Link to this heading"></a></h1>
<p>This tutorial demonstrates how to integrate whole‑brain spatial transcriptomics data and structural connectivity data in mouse using SpaCon. The spatial transcriptomics data used here is the MERFISH dataset from the study published in <em>Nature</em>: <a class="reference external" href="https://www.nature.com/articles/s41586-023-06808-9">Molecularly defined and spatially resolved cell atlas of the whole mouse brain</a>. The structural connectivity data are sourced from the <a class="reference external" href="https://connectivity.brain-map.org/">Allen Mouse Brain Connectivity
Atlas</a>.</p>
<p>We have co‑registered both datasets to the same resolution. The processed data for this tutorial can be downloaded from this <a class="reference external" href="https://drive.google.com/drive/folders/1lQQQVjXt8lvciuKq_Jkp4LGsR_EQXFAi?usp=sharing">Google Drive link</a>.</p>
<blockquote>
<div><p><strong>Note:</strong> If you want to use your own data, please make sure that your transcriptomics and connectivity matrices are paired spot‑by‑spot (i.e., transcriptomics as a spot × gene matrix and connectivity as a spot × spot matrix).</p>
</div></blockquote>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import spacon
from spacon.utils import build_spatial_graph, build_connection_graph, neighbor_sample, model_train, model_eval, clustering, plot_all_results

from spacon.model import SpaCon
import datetime
import os
import scanpy as sc
import pandas as pd
import matplotlib.pyplot as plt

import torch
import numpy as np
import pickle
import random

import warnings
warnings.filterwarnings(&quot;ignore&quot;)

mus = &#39;mouse_1&#39;
if mus == &#39;mouse_1&#39;:  # coronal
    plot_x, plot_y = &#39;z&#39;, &#39;y&#39;
    figsize = (5,5)
elif mus == &#39;mouse_3&#39;:   # sagittal
    plot_x, plot_y = &#39;x&#39;, &#39;y&#39;
    figsize = (11,5)

cluster_color=[&#39;#537eb7&#39;,&#39;#83ab8e&#39;,&#39;#ece399&#39;,&#39;#405993&#39;,&#39;#cc7f73&#39;,&#39;#d69971&#39;,&#39;#df5734&#39;,&#39;#6c408e&#39;,&#39;#9a70a8&#39;,&#39;#d4c2db&#39;,&#39;#53738c&#39;,&#39;#a25087&#39;,&#39;#a78982&#39;,&#39;#a9c2cb&#39;,&#39;#92699e&#39;,&#39;#4b6aa8&#39;,&#39;#3ca0cf&#39;,&#39;#c376a7&#39;,&#39;#ad98c3&#39;,&#39;#408444&#39;,&#39;#b95055&#39;,&#39;#d5bb72&#39;,
               &#39;#bc9a7f&#39;,&#39;#e0cfda&#39;,&#39;#d8a0c0&#39;,&#39;#e6b884&#39;,&#39;#b05545&#39;,&#39;#d69a55&#39;,&#39;#64a776&#39;,&#39;#cbdaa9&#39;,&#39;#efd2c9&#39;,&#39;#da6f6d&#39;,&#39;#ebb1a4&#39;,&#39;#a44e89&#39;,&#39;#8c564b&#39;,&#39;#b85292&#39;,&#39;#6d6fa0&#39;,&#39;#8d689d&#39;,&#39;#c8c7e1&#39;,&#39;#d25774&#39;,&#39;#c49abc&#39;,&#39;#a5a9b0&#39;,&#39;#927c9a&#39;,&#39;#9f8d89&#39;,
               &#39;#72567a&#39;,&#39;#63a3b8&#39;,&#39;#c4daec&#39;,&#39;#61bada&#39;,&#39;#b7deea&#39;,&#39;#e29eaf&#39;,&#39;#4490c4&#39;,&#39;#e6e2a3&#39;,&#39;#de8b36&#39;,&#39;#c4612f&#39;,&#39;#ad98c3&#39;,&#39;#76a2be&#39;,&#39;#cea5c7&#39;,&#39;#c6adb0&#39;,&#39;#9d3b62&#39;,&#39;#2d3462&#39;, &quot;#ff420e&quot;, &quot;#ffbb00&quot;]

def set_seed(seed: int):
    os.environ[&#39;PYTHONHASHSEED&#39;] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

set_seed(42)
</pre></div>
</div>
</div>
<p><strong>Data preprocessing</strong></p>
<p><strong>Load spatial transcriptomics data</strong></p>
<p>If there are too many genes (for example, more than 5,000), we recommend first screening for highly variable genes using the following method:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>n_top_genes = 3000
sc.pp.highly_variable_genes(adata, flavor=&quot;seurat&quot;, n_top_genes=n_top_genes)
adata = adata[:, adata.var.highly_variable]
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.read_h5ad(f&#39;./data/{mus}/adata_merge.h5ad&#39;)
print(&#39;raw adata shape:&#39;, adata.shape)
sc.pp.normalize_total(adata, target_sum=1e4)
sc.pp.log1p(adata)
adata
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
raw adata shape: (250535, 1122)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 250535 × 1122
    obs: &#39;x_section_mean&#39;, &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;section&#39;, &#39;NT_index&#39;, &#39;Cells_id&#39;
    uns: &#39;log1p&#39;
</pre></div></div>
</div>
<p><strong>Build spatial graph</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_spatial_graph</span></code> function constructs a spatial graph using the three-dimensional spatial coordinates of the spatial transcriptome. The main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: Spatial transcriptomics data must include the three-dimensional coordinates for each spot (i.e., the slice number and the two-dimensional coordinates within that slice).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">section_order</span></code>: A slice order list where the sequence represents the original arrangement of each slice within the brain. While the slices can be oriented differently, their relative order must be strictly maintained.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rad_cutoff</span></code>: Neighborhood radius, each spot will have edges added to all other spots within its neighborhood radius.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rad_cutoff_Zaxis</span></code>: Inter-slice neighborhood radius, each spot will have edges added to spots in adjacent slices that are within this radius.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sec_x</span></code>: The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the x-coordinate of each spot within its slice.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sec_y</span></code>: The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the y-coordinate of each spot within its slice.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">key_section</span></code>: Column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the slice number (where different numbers indicate different slices).</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># build the section list
section_order = np.unique(adata.obs[&#39;section&#39;]).tolist()
# calculate the spatial graph for the adata
ST_graph_data, st_adj = build_spatial_graph(adata=adata, rad_cutoff=0.7, rad_cutoff_Zaxis=1.0,
                                            sec_x=&#39;z&#39;, sec_y=&#39;y&#39;, key_section=&#39;section&#39;,
                                            section_order=section_order)
ST_graph_data
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 129/129 [00:01&lt;00:00, 88.06it/s]
100%|██████████| 128/128 [00:54&lt;00:00,  2.35it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[250535, 1122], edge_index=[2, 1681795])
</pre></div></div>
</div>
<p><strong>Load connectivity data and build connection graph</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_connection_graph</span></code> function uses connection information to construct a three-dimensional connection graph. The main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">nt_adj</span></code>: An n x n two-dimensional matrix, where n is the number of spots in the spatial transcriptomics data, representing the connection strength between spots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">threshold</span></code>: Filtering threshold, connection strengths below this value will be set to zero.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># with open(f&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/ST/zxw/{mus}/PyG_Data_xy0.7_z1.0.pkl&#39;, &#39;rb&#39;) as f:
#     ST_graph_data = pickle.load(f)
# st_adj = load_npz(f&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/ST/zxw/{mus}/zxw_adj.npz&#39;)
# nt_adj = np.load(&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/NT/zxw/mouse_1/zxw_adj.npy&#39;)
# NT_graph_data = build_connection_graph(adata, nt_adj, threshold=0.001)

eps = 0.001
proportion_after = 0.0211
with open(f&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/NT/zxw/{mus}/PyG_Data_eps{eps}_{proportion_after}.pkl&#39;, &#39;rb&#39;) as f:
    NT_graph_data = pickle.load(f)

NT_graph_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[250535, 1122], edge_index=[2, 1324207857])
</pre></div></div>
</div>
<p><strong>Neighbor-based subgraph sampling</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">neighbor_sample</span></code> function performs subgraph sampling from the input spatial graph and connection graph. Its main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_size</span></code>: The batch size for model training.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">train_num_neighbors</span></code>: The number of neighbors to sample for each node in each iteration. This parameter is used by the data loader during the model training process.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">eval_num_neighbors</span></code>: The number of neighbors to sample for each node in each iteration. This parameter is used by the data loader during the model evaluation process. If an entry is set to -1, all neighbors will be included.(default:<code class="docutils literal notranslate"><span class="pre">[-1]</span></code>)</p></li>
</ul>
<p>The function returns three data loaders: <code class="docutils literal notranslate"><span class="pre">train_loader</span></code>, <code class="docutils literal notranslate"><span class="pre">evaluate_loader_con</span></code>, and <code class="docutils literal notranslate"><span class="pre">evaluate_loader_spa</span></code>. The <code class="docutils literal notranslate"><span class="pre">train_loader</span></code> is used during the <strong>model training process</strong>. Meanwhile, <code class="docutils literal notranslate"><span class="pre">evaluate_loader_con</span></code> and <code class="docutils literal notranslate"><span class="pre">evaluate_loader_spa</span></code> are used for <strong>model evaluation</strong> on the <strong>connection graph</strong> and <strong>spatial graph</strong>, respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>train_loader, evaluate_loader_con, evaluate_loader_spa = neighbor_sample(NT_graph_data, ST_graph_data, batch_size=64, train_num_neighbors=[20, 10, 10], eval_num_neighbors=[-1], num_workers=4)
</pre></div>
</div>
</div>
<p><strong>Model training</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>device = torch.device(&#39;cuda:1&#39; if torch.cuda.is_available() else &#39;cpu&#39;)

# hyper-parameters
num_epoch = 4
lr = 0.0001
weight_decay = 1e-4
hidden_dims = [adata.X.shape[1]] + [256, 64, 16]
# model
# fusion_method indicates the feature fusion method of the middle layer, you can choose &#39;add&#39; or &#39;concat&#39;
model = SpaCon(hidden_dims=hidden_dims, fusion_method=&#39;concat&#39;).to(device)
# if model_save_path=None, the model will not be saved
results_save_path = f&quot;./results/{mus}/{str(datetime.datetime.now().strftime(&#39;%Y_%m_%d_%H_%M_%S&#39;))}/&quot;
os.makedirs(results_save_path, exist_ok=True)

model = model_train(num_epoch, lr, weight_decay, model, train_loader, st_adj, model_save_path=results_save_path, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:1|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:12&lt;00:00, 10.50it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:2|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:04&lt;00:00, 10.75it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:3|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [05:58&lt;00:00, 10.91it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:4|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:05&lt;00:00, 10.71it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>


Training completed! The model parameters have been saved to ./results/mouse_1/2025_07_23_20_37_12/model_params.pth
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p><strong>Model evaluation</strong></p>
<p>The features obtained after model dimensionality reduction, named <code class="docutils literal notranslate"><span class="pre">feature_spa</span></code> and <code class="docutils literal notranslate"><span class="pre">feature_con</span></code>, are stored in the returned <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code>. These features can be used for subsequent cluster analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = model_eval(model, adata, NT_graph_data, ST_graph_data, evaluate_loader_con, evaluate_loader_spa, st_adj, layer_eval=True, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Evaluating: 100%|██████████| 751605/751605 [05:03&lt;00:00, 2475.72it/s]
Evaluating: 100%|██████████| 751605/751605 [00:45&lt;00:00, 16534.10it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The results have been saved in adata.obsm
AnnData object with n_obs × n_vars = 250535 × 1122
    obs: &#39;x_section_mean&#39;, &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;section&#39;, &#39;NT_index&#39;, &#39;Cells_id&#39;
    uns: &#39;log1p&#39;
    obsm: &#39;feature_spa&#39;, &#39;feature_con&#39;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p><strong>Clustering</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">clustering</span></code> function performs clustering using the louvain algorithm, with the following key parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: The AnnData object obtained previously, which contains the clustering features (<code class="docutils literal notranslate"><span class="pre">feature_spa</span></code>, <code class="docutils literal notranslate"><span class="pre">feature_con</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alpha</span></code>: This parameter adjusts the contribution of local spatial information versus global connection information in the clustering results.</p>
<ul>
<li><p>When <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">1</span></code>, the clustering will incorporate more global information.</p></li>
<li><p>When <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">0</span></code>, the clustering will focus more on local information. You can set different <code class="docutils literal notranslate"><span class="pre">alpha</span></code> values based on your downstream tasks.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">adata_save_path</span></code>: The path where the results will be saved.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cluster_resolution</span></code>: The clustering resolution used during the louvain clustering process.</p></li>
</ul>
<p>The returned <code class="docutils literal notranslate"><span class="pre">path</span></code> indicates where the clustering results are saved.</p>
<hr class="docutils" />
<p>The <code class="docutils literal notranslate"><span class="pre">plot_all_results</span></code> function is used to visualize the clustering outcomes. It plots the clustering results for all slices collectively, and also generates individual plots for each cluster category. Its key parameters are:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: The AnnData object that contains the clustering results, specifically in <code class="docutils literal notranslate"><span class="pre">adata.obs['clusters']</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">path</span></code>: The directory where the generated plots will be saved.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">figsize</span></code>: The dimensions for the output plots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">plot_x</span></code>: The x-coordinate of each spot within its respective slice for plotting.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">plot_y</span></code>: The y-coordinate of each spot within its respective slice for plotting.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>def clustering(adata, alpha, adata_save_path, cluster_resolution=0.75):
    import copy
    adata = copy.deepcopy(adata)
    from sklearn.preprocessing import StandardScaler
    scaler_standard = StandardScaler()
    f_con = scaler_standard.fit_transform(adata.obsm[&#39;feature_con&#39;])
    f_spa = scaler_standard.fit_transform(adata.obsm[&#39;feature_spa&#39;])
    f_alp = (np.exp(-4 * alpha) - 1) / (np.exp(-4) - 1)
    f_add = f_alp*f_con + (1-f_alp)*f_spa
    adata.obsm[&#39;feature_add&#39;] = f_add

    sc.pp.neighbors(adata, use_rep=&#39;feature_add&#39;, n_neighbors=40)
    sc.tl.umap(adata)


    path = adata_save_path + f&#39;feature_add_weight{alpha}/Clusters_res{cluster_resolution}/&#39;
    os.makedirs(path, exist_ok=True)

    sc.tl.louvain(adata,
                resolution=cluster_resolution,
                key_added=&quot;clusters&quot;)
    sc.pl.umap(adata,color=&#39;clusters&#39;, show=False)
    plt.tight_layout()
    plt.savefig(path+f&#39;cluster_umap_res{cluster_resolution}.png&#39;)
    adata.write_h5ad(path+&#39;adata_cluster.h5ad&#39;)

    print(f&#39;The clustering results have been saved in {path}&#39;)
    print(adata)

    return adata, path
</pre></div>
</div>
</div>
<p><strong>alpha = 1</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_alpha1, path = clustering(adata, alpha=1, adata_save_path=results_save_path, cluster_resolution=0.75)

plot_all_results(adata_alpha1, path, figsize, plot_x, plot_y)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The clustering results have been saved in ./results/mouse_1/2025_07_23_20_37_12/feature_add_weight1/Clusters_res0.75/
AnnData object with n_obs × n_vars = 250535 × 1122
    obs: &#39;x_section_mean&#39;, &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;section&#39;, &#39;NT_index&#39;, &#39;Cells_id&#39;, &#39;clusters&#39;
    uns: &#39;log1p&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;clusters&#39;, &#39;clusters_colors&#39;
    obsm: &#39;feature_spa&#39;, &#39;feature_con&#39;, &#39;feature_add&#39;, &#39;X_umap&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 59/59 [03:53&lt;00:00,  3.95s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_21_2.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_21_2.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 2D result plot
adata_alpha1.obsm[&#39;spatial&#39;] = adata_alpha1.obs[[&#39;z&#39;, &#39;y&#39;]].values
adata_alpha1.obs.index = adata_alpha1.obs.index.astype(str)
adata_s = adata_alpha1[adata_alpha1.obs[&#39;section&#39;] == &#39;Zhuang-ABCA-1.081&#39;]
adata_s = adata_s[adata_s.obs[&#39;z&#39;] &lt; 56]   # Removal of part of the right brain
sc.pl.spatial(adata_s, color=&#39;clusters&#39;, spot_size=1.25, show=True, palette=cluster_color)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_22_0.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_22_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 3D result plot
from spacon.utils import result_plot_3D

result_plot_3D(adata_alpha1, highlight_section=&#39;Zhuang-ABCA-1.081&#39;, cluster_color=cluster_color)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_23_0.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_23_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sec_cluster_df = pd.DataFrame(0, index=[str(i) for i in range(len(adata_alpha1.obs[&#39;clusters&#39;].unique()))], columns=sorted(adata_alpha1.obs[&#39;section&#39;].unique()))
for sec in sorted(adata_alpha1.obs[&#39;section&#39;].unique()):
    adata_s = adata_alpha1[adata_alpha1.obs[&#39;section&#39;] == sec]
    cluster_counts = adata_s.obs[&#39;clusters&#39;].value_counts()
    for cluster, count in cluster_counts.items():
        sec_cluster_df.loc[cluster, sec] = count


sec_cluster_df_proportions = sec_cluster_df.div(sec_cluster_df.sum(axis=0), axis=1)
sec_cluster_df_proportions = sec_cluster_df_proportions.iloc[:, ::-1]
sec_cluster_df_proportions = sec_cluster_df_proportions.dropna()


from scipy.spatial.distance import jensenshannon

jsd_diff = []
for i in range(len(sec_cluster_df_proportions.columns) - 1):
    slice1 = sec_cluster_df_proportions.columns[i]
    slice2 = sec_cluster_df_proportions.columns[i+1]

    p = sec_cluster_df_proportions.iloc[:, i].values.astype(float)
    q = sec_cluster_df_proportions.iloc[:, i+1].values.astype(float)
    jsd = jensenshannon(p, q)
    jsd_diff.append(jsd)

plt.figure(figsize=(20,2))
line, = plt.plot(jsd_diff, marker=&#39;o&#39;, linestyle=&#39;-&#39;, label=&#39;JSD&#39;, lw=0.5, ms=2)

plt.axhline(0.2, color=&#39;red&#39;, linestyle=&#39;--&#39;, label=&#39;Threshold (0.2)&#39;, lw=0.5)
plt.xlabel(&#39;Adjacent Slice Pair Index&#39;)
plt.ylabel(&#39;Jensen-Shannon Divergence&#39;)
plt.legend()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x7fb670643f10&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_24_1.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_24_1.png" />
</div>
</div>
<p><strong>alpha = 0</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_alpha0, path = clustering(adata, alpha=0, adata_save_path=results_save_path, cluster_resolution=0.75)

plot_all_results(adata_alpha0, path, figsize, plot_x, plot_y)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The clustering results have been saved in ./results/mouse_1/2025_07_23_20_37_12/feature_add_weight0/Clusters_res0.75/
AnnData object with n_obs × n_vars = 250535 × 1122
    obs: &#39;x_section_mean&#39;, &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;section&#39;, &#39;NT_index&#39;, &#39;Cells_id&#39;, &#39;clusters&#39;
    uns: &#39;log1p&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;clusters&#39;, &#39;clusters_colors&#39;
    obsm: &#39;feature_spa&#39;, &#39;feature_con&#39;, &#39;feature_add&#39;, &#39;X_umap&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 25/25 [02:12&lt;00:00,  5.29s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_26_2.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_26_2.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 2D result plot

adata_alpha0.obsm[&#39;spatial&#39;] = adata_alpha0.obs[[&#39;z&#39;, &#39;y&#39;]].values
adata_alpha0.obs.index = adata_alpha0.obs.index.astype(str)
adata_s = adata_alpha0[adata_alpha0.obs[&#39;section&#39;] == &#39;Zhuang-ABCA-1.081&#39;]
adata_s = adata_s[adata_s.obs[&#39;z&#39;] &lt; 56]   # Removal of part of the right brain
sc.pl.spatial(adata_s, color=&#39;clusters&#39;, spot_size=1.25, show=True, palette=cluster_color)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_27_0.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_27_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 3D result plot
result_plot_3D(adata_alpha0, highlight_section=&#39;Zhuang-ABCA-1.081&#39;, cluster_color=cluster_color)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_28_0.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_28_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sec_cluster_df = pd.DataFrame(0, index=[str(i) for i in range(len(adata_alpha0.obs[&#39;clusters&#39;].unique()))], columns=sorted(adata_alpha0.obs[&#39;section&#39;].unique()))
for sec in sorted(adata_alpha0.obs[&#39;section&#39;].unique()):
    adata_s = adata_alpha0[adata_alpha0.obs[&#39;section&#39;] == sec]
    cluster_counts = adata_s.obs[&#39;clusters&#39;].value_counts()
    for cluster, count in cluster_counts.items():
        sec_cluster_df.loc[cluster, sec] = count


sec_cluster_df_proportions = sec_cluster_df.div(sec_cluster_df.sum(axis=0), axis=1)
sec_cluster_df_proportions = sec_cluster_df_proportions.iloc[:, ::-1]
sec_cluster_df_proportions = sec_cluster_df_proportions.dropna()


from scipy.spatial.distance import jensenshannon

jsd_diff = []
for i in range(len(sec_cluster_df_proportions.columns) - 1):
    slice1 = sec_cluster_df_proportions.columns[i]
    slice2 = sec_cluster_df_proportions.columns[i+1]

    p = sec_cluster_df_proportions.iloc[:, i].values.astype(float)
    q = sec_cluster_df_proportions.iloc[:, i+1].values.astype(float)
    jsd = jensenshannon(p, q)
    jsd_diff.append(jsd)

plt.figure(figsize=(20,2))
line, = plt.plot(jsd_diff, marker=&#39;o&#39;, linestyle=&#39;-&#39;, label=&#39;JSD&#39;, lw=0.5, ms=2)

plt.axhline(0.2, color=&#39;red&#39;, linestyle=&#39;--&#39;, label=&#39;Threshold (0.2)&#39;, lw=0.5)
plt.xlabel(&#39;Adjacent Slice Pair Index&#39;)
plt.ylabel(&#39;Jensen-Shannon Divergence&#39;)
plt.legend()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x7fb6ca5ccf10&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_29_1.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_29_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="Installation%20Guide%20for%20SpaCon.html" class="btn btn-neutral float-left" title="Installation Guide for SpaCon" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="SpaCon_tutorial2_for_mouse_celltype_classification.html" class="btn btn-neutral float-right" title="Tutorial 2: SpaCon for layer6 CT neuron subtype classification" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, quhaichao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>