

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data &mdash; SpaCon_tutorials 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=f2a433a1"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data" href="SpaCon_tutorial3_for_mouse_widefield_FC_data.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SpaCon_tutorials
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Installation%20Guide%20for%20SpaCon.html">Installation Guide for <code class="docutils literal notranslate"><span class="pre">SpaCon</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial1_for_mouse_3D_ST.html">Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial2_for_mouse_celltype_classification.html">Tutorial 2: SpaCon for layer6 CT neuron subtype classification</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial3_for_mouse_widefield_FC_data.html">Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SpaCon_tutorials</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/SpaCon_tutorial4_for_marmoset_ST_and_FC.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Tutorial-4:-SpaCon-for-marmoset-cortical-gene-expression-and-resting-state-functional-MRI-data">
<h1>Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data<a class="headerlink" href="#Tutorial-4:-SpaCon-for-marmoset-cortical-gene-expression-and-resting-state-functional-MRI-data" title="Link to this heading"></a></h1>
<p>This tutorial demonstrates how to use SpaCon to integrate marmoset gene expression and fMRI data. Both the transcriptomic and imaging datasets have already been mapped onto the marmoset cortical surface. The processed data for this tutorial can be downloaded from this <a class="reference external" href="https://drive.google.com/drive/folders/1lQQQVjXt8lvciuKq_Jkp4LGsR_EQXFAi?usp=sharing">Google Drive link</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import spacon
from spacon.utils import create_spatial_graph_from_anndata, build_connection_graph, neighbor_sample, model_train, model_eval, clustering

from spacon.model import SpaCon
import datetime
import os
import scanpy as sc
import random
import torch
import numpy as np
import h5py

import warnings
warnings.filterwarnings(&quot;ignore&quot;)
mus = &#39;mouse_3&#39;
if mus == &#39;mouse_1&#39;:  # coronal
    plot_x, plot_y = &#39;z&#39;, &#39;y&#39;
    figsize = (5,5)
elif mus == &#39;mouse_3&#39;:   # sagittal
    plot_x, plot_y = &#39;x&#39;, &#39;y&#39;
    figsize = (11,5)


def set_seed(seed: int):
    os.environ[&#39;PYTHONHASHSEED&#39;] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

set_seed(42)
</pre></div>
</div>
</div>
<p><strong>Data preprocessing</strong></p>
<p><strong>Load spatial transcriptomics data</strong></p>
<p>If there are too many genes (for example, more than 5,000), we recommend first screening for highly variable genes using the following method:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>n_top_genes = 3000
sc.pp.highly_variable_genes(adata, flavor=&quot;seurat&quot;, n_top_genes=n_top_genes)
adata = adata[:, adata.var.highly_variable]
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.read_h5ad(&#39;/mnt/Data16Tc/home/haichao/code/SpaCon/ST_FC_cluster_marmoset/data/marmoset_cortical_5klabel_gene_exp.h5ad&#39;)
sc.pp.normalize_total(adata, target_sum=1e4)
sc.pp.log1p(adata)  # add this to make the hightly_variable work

n_top_genes = 3000
sc.pp.highly_variable_genes(adata, flavor=&quot;seurat&quot;, n_top_genes=n_top_genes)
adata = adata[:, adata.var.highly_variable]
adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
View of AnnData object with n_obs × n_vars = 5000 × 3000
    obs: &#39;x&#39;, &#39;y&#39;, &#39;z&#39;
    var: &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<p><strong>Build spatial graph</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_spatial_graph</span></code> function constructs a spatial graph using the three-dimensional spatial coordinates of the spatial transcriptome. The main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: Spatial transcriptomics data must include the three-dimensional coordinates for each spot (i.e., the slice number and the two-dimensional coordinates within that slice).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">section_order</span></code>: A slice order list where the sequence represents the original arrangement of each slice within the brain. While the slices can be oriented differently, their relative order must be strictly maintained.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rad_cutoff</span></code>: Neighborhood radius, each spot will have edges added to all other spots within its neighborhood radius.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">rad_cutoff_Zaxis</span></code>: Inter-slice neighborhood radius, each spot will have edges added to spots in adjacent slices that are within this radius.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sec_x</span></code>: The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the x-coordinate of each spot within its slice.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sec_y</span></code>: The column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the y-coordinate of each spot within its slice.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">key_section</span></code>: Column name in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that stores the slice number (where different numbers indicate different slices).</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n_neighbors = 15
ST_graph_data, st_adj = create_spatial_graph_from_anndata(adata=adata,  n_neighbors=n_neighbors, spatial_key=&#39;spatial&#39;)
ST_graph_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[5000, 3000], edge_index=[2, 75000])
</pre></div></div>
</div>
<p><strong>Load connectivity data and build connection graph</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">build_connection_graph</span></code> function uses connection information to construct a three-dimensional connection graph. The main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">nt_adj</span></code>: An n x n two-dimensional matrix, where n is the number of spots in the spatial transcriptomics data, representing the connection strength between spots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">threshold</span></code>: Filtering threshold, connection strengths below this value will be set to zero.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>with h5py.File(&#39;/mnt/Data16Tc/home/haichao/code/SpaCon/ST_FC_cluster_marmoset/data/sub_all_lh_LR_bolddata_first512_5klabel_corr.mat&#39;, &#39;r&#39;) as f:
    print(list(f.keys()))
    marmoset_FC = f[&#39;mean_corr_data&#39;][()]
    print(marmoset_FC.shape)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;mean_corr_data&#39;]
(5000, 5000)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>thr = 0.25
count = np.sum(marmoset_FC &gt; thr)
total = marmoset_FC.size
ratio = count / total
ratio
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
np.float64(0.09871164)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>NT_graph_data = build_connection_graph(adata, marmoset_FC, threshold=thr)
NT_graph_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[5000, 3000], edge_index=[2, 2467802])
</pre></div></div>
</div>
<p><strong>Neighbor-based subgraph sampling</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">neighbor_sample</span></code> function performs subgraph sampling from the input spatial graph and connection graph. Its main parameters include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">batch_size</span></code>: The batch size for model training.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">train_num_neighbors</span></code>: The number of neighbors to sample for each node in each iteration. This parameter is used by the data loader during the model training process.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">eval_num_neighbors</span></code>: The number of neighbors to sample for each node in each iteration. This parameter is used by the data loader during the model evaluation process. If an entry is set to -1, all neighbors will be included.(default:<code class="docutils literal notranslate"><span class="pre">[-1]</span></code>)</p></li>
</ul>
<p>The function returns three data loaders: <code class="docutils literal notranslate"><span class="pre">train_loader</span></code>, <code class="docutils literal notranslate"><span class="pre">evaluate_loader_con</span></code>, and <code class="docutils literal notranslate"><span class="pre">evaluate_loader_spa</span></code>. The <code class="docutils literal notranslate"><span class="pre">train_loader</span></code> is used during the <strong>model training process</strong>. Meanwhile, <code class="docutils literal notranslate"><span class="pre">evaluate_loader_con</span></code> and <code class="docutils literal notranslate"><span class="pre">evaluate_loader_spa</span></code> are used for <strong>model evaluation</strong> on the <strong>connection graph</strong> and <strong>spatial graph</strong>, respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>train_loader, evaluate_loader_con, evaluate_loader_spa = neighbor_sample(NT_graph_data, ST_graph_data, batch_size=64, train_num_neighbors=[20, 10, 10], num_workers=4)
</pre></div>
</div>
</div>
<p><strong>Model training</strong></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>device = torch.device(&#39;cuda:0&#39; if torch.cuda.is_available() else &#39;cpu&#39;)

# hyper-parameters
num_epoch = 20
lr = 0.0001
weight_decay = 1e-4
hidden_dims = [adata.X.shape[1]] + [512, 128, 32]
# model
# fusion_method indicates the feature fusion method of the middle layer, you can choose &#39;add&#39; or &#39;concat&#39;
model = SpaCon(hidden_dims=hidden_dims, fusion_method=&#39;concat&#39;).to(device)
# if model_save_path=None, the model will not be saved
results_save_path = f&quot;./results_marmoset/{str(datetime.datetime.now().strftime(&#39;%Y_%m_%d_%H_%M_%S&#39;))}/&quot;
os.makedirs(results_save_path, exist_ok=True)

model = model_train(num_epoch, lr, weight_decay, model, train_loader, st_adj, model_save_path=results_save_path, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:1|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:04&lt;00:00, 18.40it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:2|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.60it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:3|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.07it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:4|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:04&lt;00:00, 19.47it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:5|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 19.94it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:6|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.03it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:7|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:04&lt;00:00, 19.54it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:8|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.30it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:9|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.17it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:10|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 19.80it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:11|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 19.83it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:12|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.12it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:13|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.34it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:14|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 19.78it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:15|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.34it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:16|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.11it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:17|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:04&lt;00:00, 19.68it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:18|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 19.94it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:19|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.03it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:20|20
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 79/79 [00:03&lt;00:00, 20.12it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>


Training completed! The model parameters have been saved to ./results_marmoset/2025_07_24_15_29_14/model_params.pth
</pre></div></div>
</div>
<p><strong>Model evaluation</strong></p>
<p>The features obtained after model dimensionality reduction, named <code class="docutils literal notranslate"><span class="pre">feature_spa</span></code> and <code class="docutils literal notranslate"><span class="pre">feature_con</span></code>, are stored in the returned <code class="docutils literal notranslate"><span class="pre">adata.obsm</span></code>. These features can be used for subsequent cluster analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = model_eval(model, adata, NT_graph_data, ST_graph_data, evaluate_loader_con, evaluate_loader_spa, st_adj, layer_eval=True, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Evaluating: 100%|██████████| 15000/15000 [00:08&lt;00:00, 1779.32it/s]
Evaluating: 100%|██████████| 15000/15000 [00:03&lt;00:00, 4768.18it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The results have been saved in adata.obsm
AnnData object with n_obs × n_vars = 5000 × 3000
    obs: &#39;x&#39;, &#39;y&#39;, &#39;z&#39;
    var: &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;
    obsm: &#39;spatial&#39;, &#39;feature_spa&#39;, &#39;feature_con&#39;
</pre></div></div>
</div>
<p><strong>Clustering</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">clustering</span></code> function performs clustering using the louvain algorithm, with the following key parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">adata</span></code>: The AnnData object obtained previously, which contains the clustering features (<code class="docutils literal notranslate"><span class="pre">feature_spa</span></code>, <code class="docutils literal notranslate"><span class="pre">feature_con</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alpha</span></code>: This parameter adjusts the contribution of local spatial information versus global connection information in the clustering results.</p>
<ul>
<li><p>When <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">1</span></code>, the clustering will incorporate more global information.</p></li>
<li><p>When <code class="docutils literal notranslate"><span class="pre">alpha</span> <span class="pre">=</span> <span class="pre">0</span></code>, the clustering will focus more on local information. You can set different <code class="docutils literal notranslate"><span class="pre">alpha</span></code> values based on your downstream tasks.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">adata_save_path</span></code>: The path where the results will be saved.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cluster_resolution</span></code>: The clustering resolution used during the louvain clustering process.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata, path = clustering(adata, alpha=1, adata_save_path=results_save_path, cluster_resolution=0.4)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The clustering results have been saved in ./results_marmoset/2025_07_24_15_29_14/feature_add_weight1/Clusters_res0.4/
AnnData object with n_obs × n_vars = 5000 × 3000
    obs: &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;clusters&#39;
    var: &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;log1p&#39;, &#39;hvg&#39;, &#39;neighbors&#39;, &#39;umap&#39;, &#39;clusters&#39;, &#39;clusters_colors&#39;
    obsm: &#39;spatial&#39;, &#39;feature_spa&#39;, &#39;feature_con&#39;, &#39;feature_add&#39;, &#39;X_umap&#39;
    obsp: &#39;distances&#39;, &#39;connectivities&#39;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial4_for_marmoset_ST_and_FC_21_1.png" src="_images/SpaCon_tutorial4_for_marmoset_ST_and_FC_21_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import nibabel as nib

# --------------------- 1. Read parcel label file ---------------------
shape_gii = nib.load(&#39;/mnt/Data16Tc/home/haichao/code/SpaCon/ST_FC_cluster_marmoset/surface/surfFS.lh.MBM_cortex_vPaxinos_5klabel.shape.gii&#39;)
parcel_labels = shape_gii.darrays[0].data  # Shape: (n_vertices,)

# --------------------- 2. Read cluster labels and convert to integer ---------------------
# Extract cluster label column and convert to integer
cluster_labels_str = adata.obs[&#39;clusters&#39;].values  # String format
cluster_labels = cluster_labels_str.astype(int)  # Convert to integer array
cluster_labels = cluster_labels+1
# cluster_labels[cluster_labels != 4] = 0

# --------------------- 3. Map parcel cluster labels to vertices ---------------------
# Initialize vertex cluster label array (background label 0 remains 0)
vertex_clusters = np.zeros_like(parcel_labels, dtype=int)

for parcel_id in range(1, 5001):  # Iterate through parcels 1-5000
    # Find vertex indices belonging to the current parcel
    mask = (parcel_labels == parcel_id)
    # Assign the current parcel&#39;s cluster label to these vertices
    vertex_clusters[mask] = cluster_labels[parcel_id - 1]  # DataFrame row index starts from 0

# --------------------- 4. Save as .func.gii file ---------------------
metadata = {
    &#39;AnatomicalStructurePrimary&#39;: &#39;CortexLeft&#39;,  # Must be consistent with the surface file
    &#39;GeometricType&#39;: &#39;Functional&#39;,  # Set to &#39;Functional&#39; (original surface file was &#39;Unknown&#39;)
    &#39;Name&#39;: &#39;ClusterLabels&#39;,
    &#39;UniqueID&#39;: &#39;CustomID&#39;,  # Custom unique ID
}
# Create GIfTI data array
data_array = nib.gifti.GiftiDataArray(
    data=vertex_clusters,
    intent=&#39;NIFTI_INTENT_NONE&#39;,  # Or &#39;NIFTI_INTENT_LABEL&#39; (depending on data type)
    datatype=&#39;NIFTI_TYPE_INT32&#39;,
    meta=nib.gifti.GiftiMetaData(metadata)  # Add metadata
)

# Create GIfTI image object and save
func_gii = nib.gifti.GiftiImage(darrays=[data_array])
nib.save(func_gii, path + &#39;cluster_labels.func.gii&#39;)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="SpaCon_tutorial3_for_mouse_widefield_FC_data.html" class="btn btn-neutral float-left" title="Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, quhaichao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>