

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity &mdash; SpaCon_tutorials 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=f2a433a1"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial 2: SpaCon for layer6 CT neuron subtype classification" href="SpaCon_tutorial2_for_mouse_celltype_classification.html" />
    <link rel="prev" title="This is SpaCon’s documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            SpaCon_tutorials
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Data-preprocessing">Data preprocessing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial2_for_mouse_celltype_classification.html">Tutorial 2: SpaCon for layer6 CT neuron subtype classification</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial3_for_mouse_widefield_FC_data.html">Tutorial 3: SpaCon for mouse spatial transcriptomics and widefield functional connectivity data</a></li>
<li class="toctree-l1"><a class="reference internal" href="SpaCon_tutorial4_for_marmoset_ST_and_FC.html">Tutorial 4: SpaCon for marmoset cortical gene expression and resting-state functional MRI data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">SpaCon_tutorials</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/SpaCon_tutorial1_for_mouse_3D_ST.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Tutorial-1:-SpaCon-for-mouse-3D-spatial-transcriptome-and-structural-connectivity">
<h1>Tutorial 1: SpaCon for mouse 3D spatial transcriptome and structural connectivity<a class="headerlink" href="#Tutorial-1:-SpaCon-for-mouse-3D-spatial-transcriptome-and-structural-connectivity" title="Link to this heading"></a></h1>
<p>This tutorial demonstrates how to integrate whole‑brain mouse spatial transcriptomics data with structural connectivity data using SpaCon. The spatial transcriptomics data are MERFISH datasets provided by Zhang et al. (2023) (available at <a class="reference external" href="https://www.nature.com/articles/s41586-023-06808-9">https://www.nature.com/articles/s41586-023-06808-9</a>), and the structural connectivity data are from the Allen Mouse Brain Connectivity Atlas (<a class="reference external" href="https://connectivity.brain">https://connectivity.brain</a>‑map.org/). Both datasets have been registered to the same resolution; the processed data can
be downloaded from XXXXX.</p>
<p>Note: To use your own data, first ensure that the spatial transcriptomics and connectivity datasets are matched one‑to‑one (the transcriptomics data should be a spot × gene matrix and the connectivity data a spot × spot matrix).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import sys
sys.path.append(&quot;/mnt/Data16Tc/home/haichao/code/SpaCon_github/SpaCon/SpaCon/&quot;)
from scipy.sparse import coo_matrix, save_npz, load_npz
from model.utils import build_spatial_graph, build_connection_graph, neighbor_sample, model_train, model_eval, clustering

from model.Model_Pyg_SpaCon import SpaCon
# from model.r_mcluster import mclust_R

import torch.nn.functional as F
import copy
import datetime
import os
import scanpy as sc
import pandas as pd
import matplotlib.pyplot as plt

import torch
import numpy as np
from tqdm import tqdm
import pickle
import random

import warnings
warnings.filterwarnings(&quot;ignore&quot;)

mus = &#39;mouse_1&#39;
if mus == &#39;mouse_1&#39;:  # coronal
    plot_x, plot_y = &#39;z&#39;, &#39;y&#39;
    figsize = (5,5)
elif mus == &#39;mouse_3&#39;:   # sagittal
    plot_x, plot_y = &#39;x&#39;, &#39;y&#39;
    figsize = (11,5)


cluster_color=[&#39;#537eb7&#39;,&#39;#83ab8e&#39;,&#39;#ece399&#39;,&#39;#405993&#39;,&#39;#cc7f73&#39;,

             &#39;#d69971&#39;,&#39;#df5734&#39;,&#39;#6c408e&#39;,&#39;#9a70a8&#39;,&#39;#d4c2db&#39;,
             &#39;#53738c&#39;,&#39;#a25087&#39;,&#39;#a78982&#39;,&#39;#a9c2cb&#39;,&#39;#92699e&#39;,
             &#39;#4b6aa8&#39;,&#39;#3ca0cf&#39;,&#39;#c376a7&#39;,&#39;#ad98c3&#39;,&#39;#408444&#39;,
             &#39;#b95055&#39;,&#39;#d5bb72&#39;,&#39;#bc9a7f&#39;,&#39;#e0cfda&#39;,&#39;#d8a0c0&#39;,
             &#39;#e6b884&#39;,&#39;#b05545&#39;,&#39;#d69a55&#39;,&#39;#64a776&#39;,&#39;#cbdaa9&#39;,
             &#39;#efd2c9&#39;,&#39;#da6f6d&#39;,&#39;#ebb1a4&#39;,&#39;#a44e89&#39;,&#39;#8c564b&#39;,
             &#39;#b85292&#39;,&#39;#6d6fa0&#39;,&#39;#8d689d&#39;,&#39;#c8c7e1&#39;,&#39;#d25774&#39;,
             &#39;#c49abc&#39;,&#39;#a5a9b0&#39;,&#39;#927c9a&#39;,&#39;#9f8d89&#39;,&#39;#72567a&#39;,
             &#39;#63a3b8&#39;,&#39;#c4daec&#39;,&#39;#61bada&#39;,&#39;#b7deea&#39;,&#39;#e29eaf&#39;,
             &#39;#4490c4&#39;,&#39;#e6e2a3&#39;,&#39;#de8b36&#39;,&#39;#c4612f&#39;,&#39;#ad98c3&#39;,
             &#39;#76a2be&#39;,&#39;#cea5c7&#39;,&#39;#c6adb0&#39;,&#39;#9d3b62&#39;,&#39;#2d3462&#39;, &quot;#FF420E&quot;, &quot;#FFBB00&quot;]

def set_seed(seed: int):
    os.environ[&#39;PYTHONHASHSEED&#39;] = str(seed)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

set_seed(42)
</pre></div>
</div>
</div>
<section id="Data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#Data-preprocessing" title="Link to this heading"></a></h2>
<p>load spatial transcriptomics data</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.read_h5ad(f&#39;./data/{mus}/adata_merge.h5ad&#39;)
print(&#39;raw adata shape:&#39;, adata.shape)
sc.pp.normalize_total(adata, target_sum=1e4)
sc.pp.log1p(adata)
adata
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
raw adata shape: (250535, 1122)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 250535 × 1122
    obs: &#39;x_section_mean&#39;, &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;section&#39;, &#39;NT_index&#39;, &#39;Cells_id&#39;
    uns: &#39;log1p&#39;
</pre></div></div>
</div>
<p>build spatial graph</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># build the section list
section_order = np.unique(adata.obs[&#39;section&#39;]).tolist()
# calculate the spatial graph for the adata
ST_graph_data, st_adj = build_spatial_graph(adata=adata,  rad_cutoff=0.7, rad_cutoff_Zaxis=1.0,
                                            sec_x=&#39;z&#39;, sec_y=&#39;y&#39;, key_section=&#39;section&#39;,
                                            section_order=section_order)
ST_graph_data
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 129/129 [00:01&lt;00:00, 80.10it/s]
100%|██████████| 128/128 [01:24&lt;00:00,  1.51it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[250535, 1122], edge_index=[2, 1681795])
</pre></div></div>
</div>
<p>load connectivity data and build connection graph</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>eps = 0.001
proportion_after = 0.0211
with open(f&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/NT/zxw/{mus}/PyG_Data_eps{eps}_{proportion_after}.pkl&#39;, &#39;rb&#39;) as f:
    NT_graph_data = pickle.load(f)

with open(f&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/ST/zxw/{mus}/PyG_Data_xy0.7_z1.0.pkl&#39;, &#39;rb&#39;) as f:
    ST_graph_data = pickle.load(f)
st_adj = load_npz(f&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/ST/zxw/{mus}/zxw_adj.npz&#39;)
# nt_adj = np.load(&#39;/mnt/Data18Td/Data/haichao/mouse_connect_data/NT/zxw/mouse_1/zxw_adj.npy&#39;)
# NT_graph_data = build_connection_graph(adata, nt_adj, threshold=0.001)
NT_graph_data
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Data(x=[250535, 1122], edge_index=[2, 1324207857])
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>train_loader, evaluate_loader_con, evaluate_loader_spa = neighbor_sample(NT_graph_data, ST_graph_data, batch_size=64, train_num_neighbors=[20, 10, 10], num_workers=4)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>device = torch.device(&#39;cuda:1&#39; if torch.cuda.is_available() else &#39;cpu&#39;)

# hyper-parameters
num_epoch = 4
lr = 0.0001
weight_decay = 1e-4
hidden_dims = [adata.X.shape[1]] + [256, 64, 16]
# model
model = SpaCon(hidden_dims=hidden_dims, fusion_method=&#39;add&#39;).to(device)
# if model_save_path=None, the model will not be saved
results_save_path = f&quot;./results/{mus}/{str(datetime.datetime.now().strftime(&#39;%Y_%m_%d_%H_%M_%S&#39;))}/&quot;
os.makedirs(results_save_path, exist_ok=True)

model = model_train(num_epoch, lr, weight_decay, model, train_loader, st_adj, model_save_path=results_save_path, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:1|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:04&lt;00:00, 10.73it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:2|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:20&lt;00:00, 10.29it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:3|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:07&lt;00:00, 10.65it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
epoch:4|4
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3915/3915 [06:20&lt;00:00, 10.30it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Training completed! The model parameters have been saved to ./results/mouse_1/2025_07_11_15_02_06//model_params.pth
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span>adata = model_eval(model, adata, NT_graph_data, ST_graph_data, evaluate_loader_con, evaluate_loader_spa, st_adj, layer_eval=True, device=device)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Evaluating: 100%|██████████| 751605/751605 [06:12&lt;00:00, 2017.67it/s]
Evaluating: 100%|██████████| 751605/751605 [01:05&lt;00:00, 11541.67it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
The results have been saved in adata.obsm
AnnData object with n_obs × n_vars = 250535 × 1122
    obs: &#39;x_section_mean&#39;, &#39;x&#39;, &#39;y&#39;, &#39;z&#39;, &#39;section&#39;, &#39;NT_index&#39;, &#39;Cells_id&#39;
    uns: &#39;log1p&#39;
    obsm: &#39;feature_spa&#39;, &#39;feature_con&#39;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.read_h5ad(&#39;/mnt/Data16Tc/home/haichao/code/SpaCon/ST_NT_cluster/SpaCon_apply_zxw/results/mouse_1/2024_04_08_20_10_00GATE_2encoder_cat_feature_decoder_skip_connect_bn/feature_add_weight1/Clusters_res0.75/adata_cluster_feature.h5ad&#39;)
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># adata = clustering(adata, alpha=1, adata_save_path=results_save_path, cluster_resolution=0.75,
#                           plot_all_cluster_results=False, figsize=figsize, plot_x=plot_x, plot_y=plot_y)
sc.pl.umap(adata,color=&#39;clusters&#39;, show=True)
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_14_0.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_14_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 2D result plot

adata.obsm[&#39;spatial&#39;] = adata.obs[[&#39;z&#39;, &#39;y&#39;]].values
adata_s = adata[adata.obs[&#39;section&#39;] == &#39;Zhuang-ABCA-1.081&#39;]
adata_s = adata_s[adata_s.obs[&#39;z&#39;] &lt; 56]   # Removal of part of the right brain
sc.pl.spatial(adata_s, color=&#39;clusters&#39;, spot_size=1.25, show=False, palette=cluster_color)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;Axes: title={&#39;center&#39;: &#39;clusters&#39;}, xlabel=&#39;spatial1&#39;, ylabel=&#39;spatial2&#39;&gt;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_15_1.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_15_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># 3D result plot

from model.utils import result_plot_3D

result_plot_3D(adata, highlight_section=&#39;Zhuang-ABCA-1.081&#39;, cluster_color=cluster_color)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_16_0.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_16_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sec_cluster_df = pd.DataFrame(0, index=[str(i) for i in range(len(adata.obs[&#39;clusters&#39;].unique()))], columns=sorted(adata.obs[&#39;section&#39;].unique()))
for sec in sorted(adata.obs[&#39;section&#39;].unique()):
    adata_s = adata[adata.obs[&#39;section&#39;] == sec]
    cluster_counts = adata_s.obs[&#39;clusters&#39;].value_counts()
    for cluster, count in cluster_counts.items():
        sec_cluster_df.loc[cluster, sec] = count


sec_cluster_df_proportions = sec_cluster_df.div(sec_cluster_df.sum(axis=0), axis=1)
sec_cluster_df_proportions = sec_cluster_df_proportions.iloc[:, ::-1]
sec_cluster_df_proportions = sec_cluster_df_proportions.dropna()


from scipy.spatial.distance import jensenshannon

jsd_diff = []
for i in range(len(sec_cluster_df_proportions.columns) - 1):
    slice1 = sec_cluster_df_proportions.columns[i]
    slice2 = sec_cluster_df_proportions.columns[i+1]

    p = sec_cluster_df_proportions.iloc[:, i].values.astype(float)
    q = sec_cluster_df_proportions.iloc[:, i+1].values.astype(float)
    jsd = jensenshannon(p, q)
    jsd_diff.append(jsd)

plt.figure(figsize=(20,2))
line, = plt.plot(jsd_diff, marker=&#39;o&#39;, linestyle=&#39;-&#39;, label=&#39;JSD&#39;, lw=0.5, ms=2)

plt.axhline(0.2, color=&#39;red&#39;, linestyle=&#39;--&#39;, label=&#39;Threshold (0.2)&#39;, lw=0.5)
plt.xlabel(&#39;Adjacent Slice Pair Index&#39;)
plt.ylabel(&#39;Jensen-Shannon Divergence&#39;)
plt.legend()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x7f4fdb6c5d90&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/SpaCon_tutorial1_for_mouse_3D_ST_17_1.png" src="_images/SpaCon_tutorial1_for_mouse_3D_ST_17_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="This is SpaCon’s documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="SpaCon_tutorial2_for_mouse_celltype_classification.html" class="btn btn-neutral float-right" title="Tutorial 2: SpaCon for layer6 CT neuron subtype classification" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, quhaichao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>